File: backend/utils/export_utils.py
"""
Utilities for exporting attendance data to Excel and PDF with formatting.

Data Format:
All exports use dict format optimized for Supabase/PostgreSQL. Each record is a dict with keys:
- id: int
- registration_no: str
- student_name: str
- course_code: str
- course_name: str
- attended_periods: int
- conducted_periods: int
- attendance_percentage: float

This format is returned by attendance_service.get_filtered_attendance_records()
and avoids ORM object overhead for better performance with remote databases.
"""
import io
import pandas as pd
from datetime import datetime
from flask import make_response
from reportlab.lib import colors
from reportlab.lib.pagesizes import A4
from reportlab.platypus import SimpleDocTemplate, Table, TableStyle, Paragraph, Spacer
from reportlab.lib.styles import getSampleStyleSheet, ParagraphStyle
from reportlab.pdfbase import pdfmetrics


class ExportUtils:
    def _timestamp_for_filename(self) -> str:
        # Example: 01.10.2025 23 05 42
        return datetime.now().strftime("%d.%m.%Y %H %M %S")

    def generate_excel_export(self, data, filter_info=None, filename_prefix: str = "attendance"):
        """Generate formatted Excel from export-ready rows (list of dict)."""
        df = pd.DataFrame(data or [])
        
        # Remove the 'id' column if present
        if 'id' in df.columns:
            df = df.drop(columns=['id'])

        output = io.BytesIO()
        with pd.ExcelWriter(output, engine='xlsxwriter') as writer:
            sheet_name = 'Low Attendance Report'
            df.to_excel(writer, sheet_name=sheet_name, index=False)

            workbook = writer.book
            worksheet = writer.sheets[sheet_name]

            header_format = workbook.add_format({
                'bold': True,
                'text_wrap': True,
                'valign': 'top',
                'fg_color': '#D7E4BC',
                'border': 1,
            })

            # Apply header styling
            for col_num, value in enumerate(df.columns.values):
                worksheet.write(0, col_num, value, header_format)

            # Auto-fit columns
            for i, col in enumerate(df.columns):
                max_len = max(df[col].astype(str).apply(len).max() if not df.empty else 0, len(str(col))) + 2
                worksheet.set_column(i, i, max_len)

        output.seek(0)

        # Build dynamic filename based on filter_info
        if filter_info:
            filter_str = " ".join(str(f).replace(":", "").replace(",", "").replace("|", "") for f in filter_info)
            filename = f"{filter_str}.xlsx"
        else:
            filename = "attendance.xlsx"
        
        # Sanitize filename to prevent header issues
        filename = filename.replace('"', '').replace("'", "")
        
        response = make_response(output.getvalue())
        response.headers['Content-Type'] = 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet'
        response.headers['Content-Disposition'] = f'attachment; filename="{filename}"'
        return response

    def generate_pdf_export(self, records, filter_info=None, filename_prefix: str = "attendance"):
        """Generate formatted PDF table for AttendanceRecord rows with wrapped cells and header-based column widths."""
        output = io.BytesIO()
        # Tighter margins to maximize usable width while staying printable
        doc = SimpleDocTemplate(
            output,
            pagesize=A4,
            leftMargin=24,
            rightMargin=24,
            topMargin=36,
            bottomMargin=36,
        )

        styles = getSampleStyleSheet()
        title_style = ParagraphStyle(
            'CustomTitle', parent=styles['Heading1'], fontSize=18, spaceAfter=7, alignment=1
        )
        filter_style = ParagraphStyle(
            'FilterInfo', parent=styles['Normal'], fontSize=11, leading=11, wordWrap='CJK'
        )
        cell_style = ParagraphStyle(
            'Cell', parent=styles['Normal'], fontSize=8, leading=10, wordWrap='LTR', alignment=1
        )

        content = []
        content.append(Paragraph("Attendance Report", title_style))
        content.append(Spacer(1, 8))

        if filter_info:
            filter_text = "Filters Applied: " + " | ".join(filter_info)
            content.append(Paragraph(filter_text, filter_style))
            content.append(Spacer(1, 8))

        headers = [
            'S.No', 'Course Code', 'Registration No', 'Student Name', 
            'Attended', 'Total', 'Attendance %'
        ]

        # Build table rows; wrap Student Name with Paragraph for proper wrapping
        table_data = [headers]
        for i, r in enumerate(records or [], 1):
            # Use dict format (optimized for Supabase/PostgreSQL)
            course_code = str(r.get('course_code', ''))
            registration_no = str(r.get('registration_no', ''))
            student_name = str(r.get('student_name', ''))
            attended = str(r.get('attended_periods', ''))
            conducted = str(r.get('conducted_periods', ''))
            percentage = r.get('attendance_percentage', 0)
            
            table_data.append([
                str(i),
                course_code,
                registration_no,
                Paragraph(student_name, cell_style),
                attended,
                conducted,
                f"{percentage:.0f}",
            ])

        # Compute column widths based on header text width + padding
        header_font = 'Helvetica-Bold'
        header_font_size = 10
        padding_lr = 12  # space on both sides

        header_min_widths = []
        for text in headers:
            w = pdfmetrics.stringWidth(text, header_font, header_font_size)
            header_min_widths.append(w + 2 * padding_lr)

        available_width = doc.width

        # Fixed columns (all except Student Name at index 3)
        fixed_indexes = [0, 1, 2, 4, 5, 6]
        fixed_sum = sum(header_min_widths[i] for i in fixed_indexes)

        # Allocate remaining width to Student Name while ensuring at least header width
        name_min = header_min_widths[3]
        name_width = max(name_min, available_width - fixed_sum)

        # If even header minimums exceed available width, cap name to its minimum
        # and proportionally scale other columns but never below their header minimums.
        col_widths = list(header_min_widths)
        col_widths[3] = name_width
        total_width = sum(col_widths)
        if total_width > available_width:
            # Compute scaling factor but clamp to not go below header minimums.
            # Start by setting all to their header minimums, then if still too wide,
            # slightly reduce numeric columns font size effect by trimming a tiny padding.
            # This preserves the requirement for at least header text width plus some space.
            # We will reduce padding on non-name columns if needed.
            excess = total_width - available_width
            adjustable_indexes = [i for i in fixed_indexes if i != 1]  # leave reg no mostly intact
            # Try to shave up to 4 points of padding from each adjustable column
            shave_per_col = min(4, excess / max(1, len(adjustable_indexes)))
            for i in adjustable_indexes:
                col_widths[i] = max(header_min_widths[i] - shave_per_col, header_min_widths[i] - 4)
            total_width = sum(col_widths)
            if total_width > available_width:
                # As a last resort, shave a bit from Registration No padding as well
                i = 1
                can_shave = min(4, total_width - available_width)
                col_widths[i] = max(header_min_widths[i] - can_shave, header_min_widths[i] - 4)

        table = Table(table_data, colWidths=col_widths)
        table_style = TableStyle([
            ('BACKGROUND', (0, 0), (-1, 0), colors.grey),
            ('TEXTCOLOR', (0, 0), (-1, 0), colors.whitesmoke),
            ('ALIGN', (0, 0), (-1, -1), 'CENTER'),
            ('ALIGN', (3, 1), (3, -1), 'CENTER'),  # Student Name center aligned
            ('VALIGN', (0, 0), (-1, -1), 'TOP'),
            ('FONTNAME', (0, 0), (-1, 0), 'Helvetica-Bold'),
            ('FONTSIZE', (0, 0), (-1, 0), 10),
            ('BOTTOMPADDING', (0, 0), (-1, 0), 7),
            ('BACKGROUND', (0, 1), (-1, -1), colors.beige),
            ('FONTNAME', (0, 1), (-1, -1), 'Helvetica'),
            ('FONTSIZE', (0, 1), (-1, -1), 8),
            ('GRID', (0, 0), (-1, -1), 1, colors.black),
        ])

        # Row highlighting (dict format - Supabase optimized)
        for i, r in enumerate(records or [], 1):
            row_index = i
            percentage = r.get('attendance_percentage', 0)
            
            if percentage < 65:
                table_style.add('BACKGROUND', (0, row_index), (-1, row_index), colors.lightcoral)
            elif percentage < 75:
                table_style.add('BACKGROUND', (0, row_index), (-1, row_index), colors.lightyellow)
            else:
                table_style.add('BACKGROUND', (0, row_index), (-1, row_index), colors.lightgreen)

        table.setStyle(table_style)
        content.append(table)
        content.append(Spacer(1, 16))
        content.append(Paragraph(
            f"Generated on: {datetime.now().strftime('%d-%m-%Y %H:%M:%S')} | Total Records: {len(records or [])}",
            styles['Normal']
        ))
        doc.title = "Attendance Report"
        doc.build(content)
        output.seek(0)

        # Build dynamic filename
        if filter_info:
            filter_str = " ".join(str(f).replace(":", "").replace(",", "").replace("|", "") for f in filter_info)
            filename = f"{filter_str}.pdf"
        else:
            filename = "attendance.pdf"
        
        # Sanitize filename to prevent header issues
        filename = filename.replace('"', '').replace("'", "")
        
        response = make_response(output.getvalue())
        response.headers['Content-Type'] = 'application/pdf'
        response.headers['Content-Disposition'] = f'attachment; filename="{filename}"'
        return response

File: backend/src/utils/exportUtils.ts
import ExcelJS from 'exceljs';
import PDFDocument from 'pdfkit';
/**
 * ExportUtils - TypeScript port of backend/utils/export_utils.py
 * Handles Excel and PDF export generation with formatting
 */
import type { Response } from 'express';

interface ExportRecord {
  id: number;
  registration_no: string;
  student_name: string;
  course_code: string;
  course_name: string;
  attended_periods: number;
  conducted_periods: number;
  attendance_percentage: number;
}

export class ExportUtils {
  /**
   * Generate timestamp for filename
   */
  private timestampForFilename(): string {
    const now = new Date();
    return now.toLocaleString('en-GB', {
      day: '2-digit',
      month: '2-digit',
      year: 'numeric',
      hour: '2-digit',
      minute: '2-digit',
      second: '2-digit'
    }).replace(/[/:]/g, ' ');
  }

  /**
   * Generate formatted Excel export
   */
  async generateExcelExport(
    data: ExportRecord[],
    filterInfo?: string[],
    filenamePrefix = 'attendance'
  ): Promise<{ buffer: Buffer; filename: string }> {
    const workbook = new ExcelJS.Workbook();
    const worksheet = workbook.addWorksheet('Low Attendance Report');

    // Define columns
    worksheet.columns = [
      { header: 'S.No', key: 'sno', width: 8 },
      { header: 'Registration No', key: 'registration_no', width: 18 },
      { header: 'Student Name', key: 'student_name', width: 30 },
      { header: 'Course Code', key: 'course_code', width: 15 },
      { header: 'Course Name', key: 'course_name', width: 40 },
      { header: 'Attended Periods', key: 'attended_periods', width: 18 },
      { header: 'Conducted Periods', key: 'conducted_periods', width: 18 },
      { header: 'Attendance %', key: 'attendance_percentage', width: 15 }
    ];

    // Style header row
    worksheet.getRow(1).font = { bold: true };
    worksheet.getRow(1).fill = {
      type: 'pattern',
      pattern: 'solid',
      fgColor: { argb: 'FFD7E4BC' }
    };
    worksheet.getRow(1).border = {
      top: { style: 'thin' },
      left: { style: 'thin' },
      bottom: { style: 'thin' },
      right: { style: 'thin' }
    };

    // Add data rows
    data.forEach((record, index) => {
      worksheet.addRow({
        sno: index + 1,
        registration_no: record.registration_no,
        student_name: record.student_name,
        course_code: record.course_code,
        course_name: record.course_name,
        attended_periods: record.attended_periods,
        conducted_periods: record.conducted_periods,
        attendance_percentage: Number(record.attendance_percentage.toFixed(1))
      });
    });

    // Add filter info at bottom
    if (filterInfo && filterInfo.length > 0) {
      worksheet.addRow([]);
      worksheet.addRow(['Filters', filterInfo.join(' | ')]);
    }

    // Generate buffer
    const buffer = await workbook.xlsx.writeBuffer();

    // Build filename
    let filename = 'attendance.xlsx';
    if (filterInfo && filterInfo.length > 0) {
      const filterStr = filterInfo
        .join(' ')
        .replace(/[:,'"|]/g, '')
        .trim();
      filename = `${filterStr}.xlsx`;
    }

    return { buffer: Buffer.from(buffer), filename };
  }

  /**
   * Generate formatted PDF export
   */
  async generatePdfExport(
    records: ExportRecord[],
    filterInfo?: string[]
  ): Promise<{ buffer: Buffer; filename: string }> {
    return new Promise((resolve, reject) => {
      const doc = new PDFDocument({ size: 'A4', margins: { left: 24, right: 24, top: 36, bottom: 36 } });
      const chunks: Buffer[] = [];

      doc.on('data', (chunk: Buffer) => chunks.push(chunk));
      doc.on('end', () => {
        const buffer = Buffer.concat(chunks);
        let filename = 'attendance.pdf';
        if (filterInfo && filterInfo.length > 0) {
          const filterStr = filterInfo.join(' ').replace(/[:,'"|]/g, '').trim();
          filename = `${filterStr}.pdf`;
        }
        resolve({ buffer, filename });
      });
      doc.on('error', reject);

      // Title
      doc.font('Helvetica-Bold').fontSize(18).text('Attendance Report', { align: 'center' });
      doc.moveDown(0.7);

      // Filter info
      if (filterInfo && filterInfo.length > 0) {
        doc.font('Helvetica').fontSize(11).text(`  Filters Applied: ${filterInfo.join(' | ')}`);
        doc.moveDown(0.4);
      }

      // Headers
      const headers = ['S.No', 'Course Code', 'Registration No', 'Student Name', 'Attended', 'Total', 'Attendance %'];
      const pageWidth = doc.page.width - doc.page.margins.left - doc.page.margins.right;
      const paddingLR = 12;
      const headerFont = 'Helvetica-Bold';
      const headerSize = 10;

      // pdfkit does not take font in options for widthOfString; set font first
      doc.font(headerFont).fontSize(headerSize);
      const headerMinWidths = headers.map(h => doc.widthOfString(h) + 2 * paddingLR);
      const fixedIdx = [0, 1, 2, 4, 5, 6];
      const fixedSum = fixedIdx.reduce((sum, i) => sum + headerMinWidths[i], 0);
      const nameMin = headerMinWidths[3];
      let nameWidth = Math.max(nameMin, pageWidth - fixedSum);

      let colWidths = [...headerMinWidths];
      colWidths[3] = nameWidth;
      let totalWidth = colWidths.reduce((a, b) => a + b, 0);
      if (totalWidth > pageWidth) {
        const adjustable = fixedIdx.filter(i => i !== 2); // keep Registration No mostly intact
        const excess = totalWidth - pageWidth;
        const shavePerCol = Math.min(4, excess / Math.max(1, adjustable.length));
        adjustable.forEach(i => { colWidths[i] = Math.max(headerMinWidths[i] - shavePerCol, headerMinWidths[i] - 4); });
        totalWidth = colWidths.reduce((a, b) => a + b, 0);
        if (totalWidth > pageWidth) {
          const i = 2; // Registration No
          const canShave = Math.min(4, totalWidth - pageWidth);
          colWidths[i] = Math.max(headerMinWidths[i] - canShave, headerMinWidths[i] - 4);
        }
      }

      // Table rendering with header as first row
      let y = doc.y;
      const rowHeight = 20; // Consistent row height matching Python
      const bodyFont = 'Helvetica';
      const bodySize = 8;

      const drawRow = (row: any[], isHeader: boolean, highlight: 'low' | 'mid' | 'high' | null) => {
        let cx = doc.page.margins.left;
        let cy = y;
        
        // Background color
        let bgColor: string;
        if (isHeader) {
          bgColor = '#808080'; // grey for header
        } else {
          bgColor = highlight === 'low' ? '#F08080' : highlight === 'mid' ? '#FFFFE0' : highlight === 'high' ? '#90EE90' : '#F5F5DC';
        }
        
        // Draw background
        doc.save();
        doc.rect(cx, cy, pageWidth, rowHeight).fill(bgColor);
        doc.restore();
        
        // Set font for row
        if (isHeader) {
          doc.font(headerFont).fontSize(headerSize).fillColor('#FFFFFF');
        } else {
          doc.font(bodyFont).fontSize(bodySize).fillColor('#000000');
        }
        
        // Draw text cells
        row.forEach((text, i) => {
          const w = colWidths[i];
          doc.text(String(text ?? ''), cx + 4, cy + 5, { width: w - 8, align: 'center' });
          cx += w;
        });
        
        // Draw grid lines
        doc.save();
        doc.strokeColor('#000000').lineWidth(1);
        let gx = doc.page.margins.left;
        doc.rect(gx, cy, pageWidth, rowHeight).stroke();
        colWidths.forEach(w => { 
          doc.moveTo(gx + w, cy).lineTo(gx + w, cy + rowHeight).stroke(); 
          gx += w; 
        });
        doc.restore();
        
        y += rowHeight;
        if (y > doc.page.height - doc.page.margins.bottom - rowHeight) {
          doc.addPage();
          y = doc.page.margins.top;
        }
      };

      // Draw header row first
      drawRow(headers, true, null);

      const rowsOut = (records || []).map((r, i) => {
        const pct = Math.round(r.attendance_percentage || 0);
        const highlight = pct < 65 ? 'low' : pct < 75 ? 'mid' : 'high';
        return {
          row: [
            String(i + 1),
            String(r.course_code ?? ''),
            String(r.registration_no ?? ''),
            String(r.student_name ?? ''),
            String(r.attended_periods ?? ''),
            String(r.conducted_periods ?? ''),
            String(pct)
          ],
          highlight
        };
      });

      rowsOut.forEach(r => drawRow(r.row, false, r.highlight as 'low' | 'mid' | 'high' | null));

      // Footer
      doc.moveDown(2);
      const generated = new Date();
      const pad = (n: number) => n.toString().padStart(2, '0');
      const ts = `${pad(generated.getDate())}-${pad(generated.getMonth() + 1)}-${generated.getFullYear()} ${pad(generated.getHours())}:${pad(generated.getMinutes())}:${pad(generated.getSeconds())}`;
      // Left-aligned footer
      doc.font('Helvetica').fontSize(10).fillColor('#000000')
        .text(`  Generated on: ${ts} | Total Records: ${records?.length ?? 0}`, doc.page.margins.left, doc.y, { align: 'left' });

      doc.end();
    });
  }

  /**
   * Send Excel export as HTTP response
   */
  sendExcelResponse(res: Response, buffer: Buffer, filename: string): void {
    res.setHeader('Content-Type', 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet');
    res.setHeader('Content-Disposition', `attachment; filename="${filename}"`);
    res.send(buffer);
  }

  /**
   * Send PDF export as HTTP response
   */
  sendPdfResponse(res: Response, buffer: Buffer, filename: string): void {
    res.setHeader('Content-Type', 'application/pdf');
    res.setHeader('Content-Disposition', `attachment; filename="${filename}"`);
    res.send(buffer);
  }
}
